README.md
# Ryzen controller

*_I'm looking for maintainers and reviewers, if you know Electron/React and have some spare times, please open an issue or ping `@storm1er#0376` or `@JamesCJ#2020` on [discord](https://discord.gg/M3hVqnT4pQ)._*

Thanks to https://github.com/FlyGoat/RyzenAdj and his author, ryzenadj.exe is included (windows only for now).

- [What's this?](#whats-this)
- [Installation](#installation)
  - [Windows](#windows)
  - [Debian like](#debian-like)
  - [Redhat like](#redhat-like)
- [Troubleshoot, Q&A](#troubleshoot-qa)
- [Development](#development)
  - [Pre-requisite](#pre-requisite)
  - [Dev install](#dev-install)
  - [Building binaries and installers](#building-binaries-and-installers)
    - [Using Windows](#using-windows)
    - [Using Linux](#using-linux)

## Disclaimers & Cautions

- If you intend to use Ryzen Controller in a video/text post online (e.g. YouTube, Reddit) you must credit the Ryzen Controller team by linking to the Ryzen Controller website! We ask this so that viewers/readers can download the software from a trusted source and so the developers get the proper recognition for their work.
- Ryzen Controller Team is not liable for any damages that my occur from using Ryzen Controller or RyzenADJ, Please use at your own risk!
- If you wish to join us developing Ryzen Controller, please join the community [discord server](https://discord.gg/bFSP5C7Awf) and ping either  `@storm1er#0376` or `@TSGames#4130`.

## What's this?

- It's a little Ryzen Master for laptops.
- Works best on 2xxx and 3xxx Ryzen series (4xxx series is experimental)

## Installation

### Windows

- Go to [release page](https://gitlab.com/ryzen-controller-team/ryzen-controller/-/releases)
- Download the latest `Ryzen Controller Setup X.X.X.exe`
- Enjoy!

### Debian like

- Go to [RyzenAdj](https://github.com/FlyGoat/RyzenAdj) repo.
- Download and build as explained in [Build requirements](https://github.com/FlyGoat/RyzenAdj#build-requirements).
- Go to Ryzen Controller's [release page](https://gitlab.com/ryzen-controller-team/ryzen-controller/-/releases)
- Download the latest `ryzen-controller_X.X.X_ARCH.deb` file.
- `sudo dpkg -i ryzen-controller_X.X.X_ARCH.deb`
- Launch with `sudo ryzencontroller`
- Set the path to your freshly builded `ryzenadj` binary into the "settings" tab.
- Enjoy!

### Redhat like

- Go to [RyzenAdj](https://github.com/FlyGoat/RyzenAdj) repo.
- Download and build as explained in [Build requirements](https://github.com/FlyGoat/RyzenAdj#build-requirements).
- Go to Ryzen Controller's [release page](https://gitlab.com/ryzen-controller-team/ryzen-controller/-/releases)
- Download the latest `ryzen-controller-X.X.X.ARCH.rpm` file.
- `sudo rpm -u ryzen-controller-X.X.X.ARCH.rpm`
- Launch with `sudo ryzencontroller`
- Set the path to your freshly builded `ryzenadj` binary into the "settings" tab.
- Enjoy!

## Troubleshoot, Q&A

> I'm getting an error when installing ryzen controller on linux

_You may need to install `smartmontools` & `lm-sensors` packages to allow Ryzen Controller to work well._
```bash
# Install the app
sudo dpkg -i ryzen-controller_x.x.x_amd64.deb
# If you get error about missing dependencies
sudo apt-get -f install
# To ensure correct temperature and others sys-info
sudo apt-get -y smartmontools lm-sensors
```

> Why yarn?

_See https://github.com/electron-userland/electron-builder/issues/1147#issuecomment-276284477_



## Development

**THIS PART IS ONLY FOR DEVELOPMENT PURPOSE, IF YOU JUST WANT TO USE RYZEN CONTROLLER, SEE THE [INSTALLATION](#installation) PART.**

### Pre-requisite

- NodeJS v10.18.0 or newer.
- About building dependencies:
  - No dependencies for windows installer
  - See [electron-installer-debian requirements](https://github.com/electron-userland/electron-installer-debian#requirements)
  - See [electron-installer-redhat requirements](https://github.com/electron-userland/electron-installer-redhat#requirements)

### Dev install

```bash
cd project
yarn install --frozen-lockfile # Please commit any change in yarn.lock/package.json in separated merge request
yarn start # You may want to look at "start:*" scripts in package.json
```

### Building binaries and installers

#### Using Windows

```bash
cd project
yarn docker # may not be needed, depends on your machine
yarn clean # You may want to look at "clean:*" scripts in package.json
yarn dist-pack-win # You may want to look at "dist-pack:*" scripts in package.json
```

#### Using Linux

```bash
cd project
yarn docker # may not be needed, depends on your machine
yarn clean # You may want to look at "clean:*" scripts in package.json
yarn dist-pack-linux # You may want to look at "dist-pack:*" scripts in package.json
```


CHANGELOG.md
# CHANGELOG

<!--- next entry here -->

## 2.6.0
2022-10-16

### Features

- update RyzenAdj to version 0.11.1 (f8fb275f4ed2e3feaab23fb885fddd92ef9d506d)

## 2.5.4
2021-10-25

### Fixes

- #180 Change autostart method to VBS script. (8afdfac3f22826d8204c8d7d806af2d393293938)

## 2.5.3
2021-09-19

### Fixes

- Better status scene, now displaying error messages. (06c9d0d7c65954d123fbd0a8e3faa6d5b4f985ff)

## 2.5.2
2021-09-12

### Fixes

- You can now set the temp. limit up to 105°C. (62158ddbf19075192e472676ace3068ac21ee46e)

## 2.5.1
2021-09-05

### Fixes

- Fix #172 Allow higher slow-time values. (58314fa8c836f52d5e60e9a9b59e271d79375c03)

## 2.5.0
2021-08-30

### Features

- You can now see the current values in the status tab (by @TSGames) (98383a6a76aaeb9de43bc73082055d8bc2fe6ebb)

### Fixes

- Updated RyzenAdj to version 0.8.2. (1b369ef54cd8182654a6e81a59f5b1f5b99c06a9)
- Updated some visual aspects for better understanding. (117c52acbb6b4122a9fa4619f23c7917aac257d3)
- Updated discord link. (570d01cdf7f83cffb89c7145530a2401a502c492)

## 2.4.1
2021-04-29

### Fixes

- Updated temp limit based on new advice (1693d277502a14840a5e0136cfbdf18c41a36f40)

## 2.4.0
2021-04-04

### Features

- Added Korean Translation, by @lyttiya. (b623fc377701efb58e2fd535660be996ee77bc75)

### Fixes

- Ensure all releases are kept forever. (3b40fe49e64123372ef12af8490d816ab6e21445)

## 2.3.0
2021-03-26

### Features

- 5000 series support (1e8364f5888dc0489c7f7c0c386bc82671490806)

## 2.2.7
2020-11-01

### Fixes

- updated links (806165e826754a0a72d6700500ebdcce1402cea5)
- updated links (d2216f4b9fb4a302141d1a6b8948deb3eebbf561)
- further alterations for Renoir (7a8aa3957fd265f38092edfc377f6ef21dfc7eb8)

## 2.2.6
2020-10-23

### Fixes

- updated readme (78b9cb52c02e7bbf49763261eccc2d123eccf893)
- further alterations of restrictions for Renoir (a4432b5ad1157cb63103e20c719db436a6154c84)

## 2.2.5
2020-09-06

### Fixes

- altered restrictions for Renoir (d7a5fc805537ae602799632c3beec2ee0d8b4627)
- update release page link (22432dbfd4c645c002e97d3b681e165e702c4cbd)

## 2.2.4
2020-08-26

### Fixes

- Updated default Slow Boost Duration to fix preset invalid error (5368ca483212095a1addf5d0a7d92a4351ed81b4)

## 2.2.3
2020-08-26

### Fixes

- fixes 3000/2000 series CPU clock clock (2d14520df22e93a0d4868209e4b02f094bb184e5)
- fixes 3000/2000 series CPU clock clock (981f916ba2bb27675bea2315a1c459bccff0bfaf)
- fixes 3000/2000 series CPU clock clock (a6c374d7ee241306e6c3205f758ac22d7b9e3b73)

## 2.2.2
2020-08-26

### Fixes

- changed settings names (b6c62d4a3f33b3fef22c1334029a0d93e3f31cf2)
- changed settings name to better illustrate what they do (8cf4af4977293ebe51256dcd0f9ef1041db8b5b9)
- fixed spelling of short boost (04f3f7513d4ec4df4cb63d139cda2781191c3297)
- fixed short duration (9ab8b5b6659bc237ba9452d4477958cf23ea930d)
- updated short and long boost limits (507cb8c75f799512f95a184dee3512d57e2319ca)
- read me file (4f1637175695f233db90fb4c527a502ce06db281)
- read me file (aa395666fcedab3011bbfbcbd8f67eb261ac98aa)
- updated read me file (0de550ae83bdf340a80b2d1f0de8a4c1e32d5216)
- 3000/2000 series fix (558c15b9186d8e4814331923919e777624dab289)
- 3000/2000 series fix (990dbefac4ef6496fcb60a542a24c1d4a75fa446)
- 3000/2000 series fix (1f8e65be339148752afe3228e2e12cc00032ce9e)
- 3000/2000 series fix (fe5e1418160747400e508b33bb11d7b927f3e052)

## 2.2.1
2020-08-23

### Fixes

- Support for Renoir APUs. (40c842c2c5e736f498b895a8fc7c12ca574d9a03)

## 2.2.0
2020-03-29

### Features

- Added Russian Translation. (f78e3c52b850ee8eb99e475db0a94d22457be93d)
- added Ukrainian lng, declared configs (f8134daae37204bec238919fab0e34470616859c)
- Interface re-styling by @danielderbenev. (c54ed5aedf531cc5b1ba2546147c42a03f134e4e)

### Fixes

- The app was unable to load the settings in some case. (6b600ccc382ee593638ffd6365e9a3bf406de479)

## 2.1.2
2020-03-21

### Fixes

- Ensure storage upgrade path from any version to above. (2388d4b97429784ee5a5f5f69c606e71719ef520)
- Unable to initialize persistent data in some cases. (2cf1007b5d82def2af6b712e2ffe3846b6dcb71f)
- #74 Reduce freezes due to translation. (df2d7d6357d9a9995b8c0f326b251d5bcc9681b7)
- Unable to change language while loading system signature in some case. (347b05bad7da0615fc887aff445d60cb05bba176)
- Auto apply presets wasn't working as expected in some cases. (5629f8034ae41d924d14887cdd28273934565a62)
- #74 Better DA and TR locales. (70989bf77e7af8d128531673b8b03a600f4cc44a)

## 2.1.1
2020-03-17

### Fixes

- The app was unable to start in some case. (1eaeb29730d532be9c5604690d0cce967a6c5847)

## 2.1.0
2020-03-17

### Features

- Display a preset compatibility note for each online preset. (775ac86f3500ba5f95ab3907cf4b13350cb1fc71)
- Added Spanish translation. (8dd94fd1261b05ca8ff4cdb45fc547318556f7c2)

### Fixes


js/methods.js
/**
 * Will enable Sentry.
 */
function addSentry() {
  const uuidv4 = require('uuid/v4');
  const settings = require('electron-settings');

  var waitSentry = setInterval(() => {
    if (!Sentry) {
      console.log('waitSentry');
      return;
    }
    clearInterval(waitSentry);
    console.log('sentryOk');
    Sentry.init({
      dsn: 'https://f80fd3ea297141a8bdc04ce812762f39@sentry.io/1513427',
      release: require('./package.json').version,
      beforeSend: (event) => {
        event.exception.values = event.exception.values.map((value) => {
          if (value.stacktrace) {
            value.stacktrace.frames = value.stacktrace.frames.map((frame) => {
              frame.filename = frame.filename.replace(/^.*ryzen(|-)controller\//g, "");
              return frame;
            });
          }
          return value;
        });
        if (event.request.url) {
          event.request.url = event.request.url.replace(/^.*ryzen(|-)controller\//g, "");
        }
        return event;
      }
    });
    Sentry.configureScope((scope) => {
      if (!settings.get('userid')) {
        settings.set('userid', uuidv4());
      }
      const userid = settings.get('userid');
      scope.setUser({
        id: userid
      });
    });
  }, 100);
}

/**
 * Will load options_data.json and display them into index.html.
 */
function displayOptions(){
  const options_data = require('./js/options_data.json');
  var tabs = {};
  for (const option_name in options_data) {
    if (!options_data.hasOwnProperty(option_name)) {
      appendLog(`Error while loading ${option_name} option.`);
      continue;
    }
    let option_data = options_data[option_name];
    if (!option_data.hasOwnProperty('tab')) {
      appendLog(`Error while loading ${option_name} tab property.`);
      continue;
    }
    let tab_name = option_data['tab'];
    let tab_name_css = tab_name.toLowerCase().replace(/ /g, "-");
    let tab = document.querySelector(`#${tab_name_css}-tab`);
    if (!tab) {
      let selectorContent = document.querySelector(`#main-container-selector`).innerHTML;
      let tabContent = document.querySelector(`#main-container`).innerHTML;

      document.querySelector('#main-container-selector').innerHTML = /*html*/`
        <li><a href="#">${tab_name}</a></li>
        ${selectorContent}
      `;
      document.querySelector(`#main-container`).innerHTML = /*html*/`
        <li class="uk-margin-top uk-margin-bottom uk-container" id="${tab_name_css}-tab"></li>
        ${tabContent}
      `;
      tab = document.querySelector(`#${tab_name_css}-tab`);
    }
    tabs[tab_name_css] = true;
    tab.innerHTML += /*html*/`
      <h3 id="${option_name}-label" uk-tooltip="${option_data.description}"><label>
        <input class="uk-checkbox uk-margin-small-right" type="checkbox" id="apply_${option_name}"/>
        <span class="option-label">${option_data.label}</span>
      </label></h3>
      <div class="uk-grid-small" uk-grid>
        <div class="uk-width-1-6">
          <input
            class="uk-input"
            type="number"
            id="${option_name}"
            repeat="${option_name}_range"
            min="${option_data.min}"
            max="${option_data.max}"
            step="${option_data.step}"
            value="${option_data.default}"
          />
        </div>
        <div class="uk-width-expand">
          <input
            class="uk-range"
            type="range"
            repeat="${option_name}"
            id="${option_name}_range"
            min="${option_data.min}"
            max="${option_data.max}"
            step="${option_data.step}"
            value="${option_data.default}"
          />
        </div>
      </div>
    `;
    UIkit.tooltip(document.querySelector(`#${option_name}-label`));
  }

  for (const tab_name_css in tabs) {
    if (tabs.hasOwnProperty(tab_name_css)) {
      const tab = document.querySelector(`#${tab_name_css}-tab`);
      tab.innerHTML += /*html*/`
        <p class="uk-margin">
          <button class="uk-button uk-button-primary" onClick="applyRyzenSettings()">Apply</button>
          <button class="uk-button uk-button-secondary" uk-toggle="target: #modal-new-preset">Save to preset</button>
        </p>
      `;
    }
  }
}

/**
 * Will create a nodes from an html string.
 * @param {string} str An html string
 */
function parseHTML(str) {
  var tmp = document.implementation.createHTMLDocument();
  tmp.body.innerHTML = str;
  return tmp.body.children;
};

/**
 * Return the current working directory.
 */
function getCurrentWorkingDirectory() {
  const cwd = require('electron').remote.app.getAppPath()
  appendLog(`getCurrentWorkingDirectory(): ${cwd}`);
  return cwd;
}

/**
 * Conversion from int to hex.
 * @param {int} number A number.
 */
function decimalToHexString(number) {
  if (number < 0)
  {
    number = 0xFFFFFFFF + number + 1;
  }

  return number.toString(16).toUpperCase();
}

/**
 * Will execute the given callback once document is ready.
 * @param {function} fn A callback to be executed.
 */
function ready(fn) {
  if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading"){
    fn();
  } else {
    document.addEventListener('DOMContentLoaded', fn);
  }
}

/**
 * Will make sure inputs are repeated when needed.
 *
 * Use the "repeat" html attribute to define where the current value must be repeated.
 */
function registerRepeaterForAllInput() {
  var ranges = document.querySelectorAll('[repeat]');
  for (const range in ranges) {
    if (ranges.hasOwnProperty(range)) {
      const element = ranges[range];
      element.addEventListener('change', function(event) {
        var repeater = document.getElementById(event.target.attributes.repeat.value);
        repeater.value = event.target.value;
      });
    }
  }
}

/**
 * Check that the app is running with admin right.
 *
 * Will display a warning if not.
 */
function checkForAdminRights() {
  if (!isWindows()) {
    const isRoot = process.getuid && process.getuid() === 0;
    if (!isRoot) {
      notification('danger',
      `Warning: you must launch this app as administrator.<br/>`
      + `Use "sudo ryzencontroller" from terminal to fix this.`
      );
    }
  } else {
    const child = require('child_process').execFile;
    child('NET SESSION', function(err,so,se) {
      if (se.length !== 0) {
        notification('warning',
          `Warning: you should launch this app as administrator, `
        + `ryzenadj.exe doesn't seems to work correctly without administrator rights.`
        );
      }
    });
  }
}

/**
 * Check that the app is running on Windows.
 */
function isWindows() {
  return require('os').platform() === 'win32';
}

/**
 * Will display a notification in ".notification-zone".
 * @param {string} type "primary", "warning", "danger" or "success".
 * @param {string} message The message to be displayed, new line will be replaced by <br/>.
 */
function notification(type, message) {
  appendLog(`notification(): ${type}\n${message}`);

  if (window.last_notification) {
    window.last_notification.close();
    window.last_notification = false;
  }
  window.last_notification = new Notification('Ryzen Controller', {
    body: message,
    requireInteraction: false,
    silent: true,
  });

  if (!window.last_notification) {
    window.last_notification = false;
    UIkit.notification({
      message: (''+message).replace(/(?:\r\n|\r|\n)/g, '<br/>'),
      status: type,
      pos: 'top-right',
      timeout: 5000,
    });
  }
}

/**
 * Will return the ryzenadj.exe path registered, or default one if not provided.
 */
function getRyzenAdjExecutablePath() {
  const settings = require('electron-settings');
  var ryzen_adj_path = settings.get('settings.ryzen_adj_path');
  if (!ryzen_adj_path && isWindows()) {
    ryzen_adj_path = getCurrentWorkingDirectory() + "\\bin\\ryzenadj.exe";
  }
  appendLog(`getRyzenAdjExecutablePath(): "${ryzen_adj_path}"`);
  return ryzen_adj_path;
}

/**
 * Will fill settings page on render with saved data.
 */
function preFillSettings() {
  const ryzen_adj_path = document.getElementById('ryzen_adj_path');
  const settings = require('electron-settings');

  ryzen_adj_path.value = getRyzenAdjExecutablePath();
  document.getElementById('start_at_boot').checked = !!settings.get('settings.start_at_boot');
  document.getElementById('apply_last_settings_on_launch').checked = !!settings.get('settings.apply_last_settings_on_launch');
  document.getElementById('minimize_to_tray').checked = !!settings.get('settings.minimize_to_tray');
  document.getElementById('start_minimized').checked = !!settings.get('settings.start_minimized');

  seconds = parseInt(settings.get('settings.reapply_periodically'));
  seconds = seconds >= 0 ? seconds : 0;
  document.getElementById('reapply_periodically').value = seconds;
  document.getElementById('reapply_periodically_range').value = seconds;
}

/**
 * Check if ryzenadj executable has been registered and is valid.
 * Will display a notification if not.
 *
 * @return {Boolean} True if ryzenadj executable exists.
 */
function isRyzenAdjPathValid() {
  var fs = require('fs');

  if (!fs.existsSync(getRyzenAdjExecutablePath())) {
    notification('danger', "Path to ryzenadj.exe is wrong, please fix it in settings tab.");
    return false;
  }
  return true;
}

/**
 * Will open a dialog to let user choose where is ryzenadj.exe.
 */
function askingForRyzenAdjExecutablePath() {
  var remote = require('electron').remote;
  var dialog = remote.require('electron').dialog;

  dialog.showOpenDialog({
    properties: ['openFile']
  }, function (filePaths) {
    const settings = require('electron-settings');
    if (typeof filePaths === 'undefined') {
      notification('warning', 'No path given, nothing changed.');
      return;
    }
    if (typeof filePaths[0] === 'undefined') {
      notification('warning', 'No path given, nothing changed.');
      return;
    }

    settings.set("settings",
      Object.assign(
        {},
        settings.get('settings'),
        { ryzen_adj_path: filePaths[0] }
      )
    );

    notification('primary', 'Path to ryzenAdj executable has been saved.');
    appendLog(`askingForRyzenAdjExecutablePath(): ${filePaths[0]}`);
    preFillSettings();
  });
}

/**
 * Will append logs to the logs tab.
 * @param {string} message The message to be logged.
 */
function appendLog(message) {
  var log_area = document.getElementById('logs');
  log_area.value += message + "\n";
  console.log(message);
}

/**
 * Will save the latest used settings.
 */
function saveLatestUsedSettings() {
  var inputs = document.querySelectorAll('#main-container *[id$="-tab"] input');
  var latest_controller_tabs_settings = {};
  inputs.forEach(element => {
    let id = element.id;
    let value = element.value;
    if (id.indexOf('apply_') === 0) {
      value = element.checked;
    }
    latest_controller_tabs_settings[id] = value;
  });
  const settings = require('electron-settings');
  let ret = settings.set("latest_controller_tabs_settings", latest_controller_tabs_settings);
  appendLog(`saveLatestUsedSettings(): ${JSON.stringify(latest_controller_tabs_settings)}`);
  appendLog(`saveLatestUsedSettings(): ${JSON.stringify(ret)}`);
}

/**
 * Will load the latest settings and refresh the controller tab's values.
 *
 * Will also apply latest settings if settings.apply_last_settings_on_launch is true.
 */
function loadLatestUsedSettings() {
  const settings = require('electron-settings');
  var latest_controller_tabs_settings = settings.get("latest_controller_tabs_settings");
  appendLog(`loadLatestUsedSettings(): ${JSON.stringify(latest_controller_tabs_settings)}`);
  for (const id in latest_controller_tabs_settings) {
    if (latest_controller_tabs_settings.hasOwnProperty(id)) {
      const value = latest_controller_tabs_settings[id];
      let input = document.getElementById(id);
      if (input) {
        if (id.indexOf('apply_') === 0) {
          input.checked = value;
        } else {
        input.value = value;
      }
    }
  }
  }
  if (document.isStarting && settings.get('settings.apply_last_settings_on_launch')) {
    applyRyzenSettings();
  }
}

/**
 * Will show or display options based on apply checkboxes value.
 */
function toggleOptionDisplayBasedOnApplyCheckbox() {
  var checkbox_toggle_options = document.querySelectorAll('#main-container *[id$="-tab"] input[id^=apply_]');
  const hideOptionBasedOnInput = function (input) {
    if (input.checked) {
      input.parentElement.parentElement.nextElementSibling.removeAttribute('hidden');
    } else {
      input.parentElement.parentElement.nextElementSibling.setAttribute('hidden', '');
    }
  }
  Array.from(checkbox_toggle_options).forEach(input => {
    hideOptionBasedOnInput(input);
    input.addEventListener('change', function(event) {
      hideOptionBasedOnInput(input);
    });
  });
}

/**
 * Listen settings tab inputs to save their values.
 */
function registerEventListenerForSettingsInput() {
  const settings = require('electron-settings');

  toggleOptionDisplayBasedOnApplyCheckbox();

  var apply_last_settings_on_launch = document.getElementById('apply_last_settings_on_launch');
  apply_last_settings_on_launch.addEventListener('change', function() {

    settings.set(
      "settings",
      Object.assign(
        {},
        settings.get('settings'),
        { apply_last_settings_on_launch: !!apply_last_settings_on_launch.checked }
      )
    );

  });
  var minimize_to_tray = document.getElementById('minimize_to_tray');
  minimize_to_tray.addEventListener('change', function() {

    settings.set(
      "settings",
      Object.assign(
        {},
        settings.get('settings'),
        { minimize_to_tray: !!minimize_to_tray.checked }
      )
    );

  });
  var start_minimized = document.getElementById('start_minimized');
  start_minimized.addEventListener('change', function() {

    settings.set(
      "settings",
      Object.assign(
        {},
        settings.get('settings'),
        { start_minimized: !!start_minimized.checked }
      )
    );

  });
  var reapply_periodically = document.getElementById('reapply_periodically');
  reapply_periodically.addEventListener('change', function() {
    reApplyPeriodically(reapply_periodically.value);

    settings.set(
      "settings",
      Object.assign(
        {},
        settings.get('settings'),
        { reapply_periodically: reapply_periodically.value }
      )
    );

  });
  var start_at_boot = document.getElementById('start_at_boot');
  start_at_boot.addEventListener('change', function() {

    settings.set(
      "settings",
      Object.assign(
        {},
        settings.get('settings'),
        { start_at_boot: !!start_at_boot.checked }
      )
    );

    updateScheduledStartOnBoot(!!start_at_boot.checked);
  });
}

/**
 * Simply display version in appropriate zone.
 */
function displayVersion() {
  const pjson = require('./package.json');
  document.getElementById('version').innerHTML = `v${pjson.version}${isDevMode() ? '-dev' : ''}`;
}

/**
 * Re-apply flow for "reapply_periodically" settings.
 * @param {number} seconds Interval in seconds between each apply.
 */
function reApplyPeriodically(seconds) {
  seconds = parseInt(seconds) >= 0 ? parseInt(seconds) : 0;
  appendLog(`reApplyPeriodically(): seconds = ${seconds}`);
  appendLog(`reApplyPeriodically(): document.reapplyLoop = ${document.reapplyLoop}`);
  clearInterval(document.reapplyLoop);
  document.reapplyLoop = false;

  if (seconds <= 0) {
    if (!document.isStarting) {
      notification('primary', "Ryzen Controller will stop applying RyzenAdj periodically.");
    }
    return;
  }

  document.reapplyLoop = setInterval(applyRyzenSettings, seconds * 1000);
}

/**
 * Will recreate shortcut on launch ... no other solution for now :(
 */
function recreateShortcut() {
  const settings = require('electron-settings');

  if (!!settings.get('settings.first_launch')) {
    let app = require('electron').remote.app;
    let fs = require('fs');
    try {
      var shortcut_path = app.getPath('desktop') + "\\Ryzen Controller";
      fs.unlink(shortcut_path, console.log);
      fs.symlink(app.getPath('exe'), shortcut_path, function (err) {
        if (err) {
          notification("danger", "Shortcut can't be created, please check log tabs for more info.");
          appendLog(`recreateShortcut(): ${err}`);
        }
        else {
          notification('primary', "A shortcut has been created on desktop.");
        }
      });
    } catch (error) {
      appendLog(`recreateShortcut() ${error}`);
      Sentry.captureException(`recreateShortcut() ${error}`);
    }
  }
}

/**
 * Will return an object completed with the current settings from inputs.
 * @param {string} keyType "inputId" or "ryzenadjArgs"
 */
function getCurrentSettings(keyType) {
  if (keyType === "ryzenadjArgs") {

    const options_data = require('./js/options_data.json');
    var settingsToBeUsed = {};
    for (const elementId in options_data) {
      if (options_data.hasOwnProperty(elementId)) {
        const optionData = options_data[elementId];
        const ryzenadjArg = optionData.ryzenadj_arg;
        if (document.getElementById('apply_' + elementId).checked) {
          settingsToBeUsed[ryzenadjArg] = document.getElementById(elementId).value;
        }
      }
    }

    appendLog('getCurrentSettings(): ' + JSON.stringify(settingsToBeUsed));
    return settingsToBeUsed;

  } else {
    var inputs = document.querySelectorAll('#main-container *[id$="-tab"] input');
    var currentSettings = {};
    inputs.forEach(element => {
      let id = element.id;
      let value = element.value;
      if (id.indexOf('apply_') === 0) {
        value = element.checked;
      }
      currentSettings[id] = value;
    });
    return currentSettings;
  }
}

/**
 * Will check for new release.
 */
function checkForNewRelease() {
  var request = new XMLHttpRequest();
  const version = require('./package.json').version;

  try {
    request.open('GET', 'https://gitlab.com/api/v4/projects/11046417/releases', true);

    request.onload = function() {
      if (this.status >= 200 && this.status < 400) {
        var resp = {};
        try {
          resp = JSON.parse(this.response);
        } catch (error) {
          console.log('WARNING: unable to check if a new version is available.', error);
          return;
        }
        if (resp[0].tag_name !== version) {
          notification('primary', "A new vesion is available, please check the release tab.");
        }
      }
    };

    request.send();
  } catch (error) {
    console.log('Unable to check if new release is available.', error);
  }
}

/**
 * Will change the BCD entry value for userplatformclock.
 *
 * @param {string} value BCD entry value for userplatformclock: "true" or "false".
 */
function toggleHpet(value) {
  value = value === "true" ? "true" : "false";
  const BcdeditExecutablePath = "C:\\Windows\\System32\\bcdedit.exe";
  const parameters = [
    "/set",
    "useplatformclock",
    value,
  ];

  const child = require('child_process').execFile;
  child(BcdeditExecutablePath, parameters, function(err, data) {
    var output = data.toString();
    if (err) {
      notification('danger', err + '<br/>' + output);
    }
    else if (output) {
      notification('success', 'Bcdedit output:<br/>' + output);
      saveLatestUsedSettings();
    }
  });
}

/**
 * Any element containing class "windows-only" will be hidden on other platform.
 */
function handlePlatformSpecificDisplay() {
  var windows_only_elements = document.getElementsByClassName('windows-only');
  if (!isWindows()) {
    for (const item of windows_only_elements) {
      item.setAttribute('hidden', 'true');
    }
  }
}

/**
 * Will delete scheduled task to start ryzen controller on session start then recreate it if isEnable is true.
 *
 * @param {bool} toBeEnabled Is auto launch should be enabled?
 */
function updateScheduledStartOnBoot(toBeEnabled) {
  const app = require('electron').remote.app;
  window.app = app;
  const AutoLaunch = require('auto-launch');
  let autoLaunch = new AutoLaunch({
    name: 'Ryzen Controller'
  });
  autoLaunch.isEnabled().then((isEnabled) => {
    console.log(`toBeEnabled: ${toBeEnabled} isEnabled: ${isEnabled}`);

    try {
      if (isEnabled) {
        autoLaunch.disable();
      }
      if (toBeEnabled) {
        autoLaunch.enable();
      }
    } catch (error) {
      console.log("WARNING: Unable to manage start on boot.", error);
    }
  });
}

/**
 * Return true if dev mode.
 */
function isDevMode() {
  return !require('electron').remote.app.isPackaged;
}

/**
 * Will listen for system events and handle status for it.
 */
function handleAcStatusChanges() {
  const powerMonitor = require('electron').remote.powerMonitor;
  const settings = require('electron-settings');

  let applyPresetOnAcStatusChange = function(presetName) {
    appendLog(`applyPresetOnAcStatusChange(${presetName})`);
    if (!presetName) {
      return;
    }
    preset_apply(presetName);
  };

  powerMonitor.on('on-ac', () => {
    applyPresetOnAcStatusChange(settings.get(`auto-apply.update-ac-plugged-in`));
  });
  powerMonitor.on('on-battery', () => {
    applyPresetOnAcStatusChange(settings.get(`auto-apply.update-ac-plugged-out`));
  });
}
